In [2]:
import pandas as pd

orders_reviews_df = pd.read_csv("data/olist_order_reviews_dataset.csv")

orders_reviews_df.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [3]:

print(orders_reviews_df.shape) #check row and column
print('\n')
print(orders_reviews_df.info()) #check dataframe information number of entries, to col names
print('\n')
print(orders_reviews_df.describe()) #check dataframe statitical summary

(100000, 7)


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 7 columns):
 #   Column                   Non-Null Count   Dtype 
---  ------                   --------------   ----- 
 0   review_id                100000 non-null  object
 1   order_id                 100000 non-null  object
 2   review_score             100000 non-null  int64 
 3   review_comment_title     11715 non-null   object
 4   review_comment_message   41753 non-null   object
 5   review_creation_date     100000 non-null  object
 6   review_answer_timestamp  100000 non-null  object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB
None


        review_score
count  100000.000000
mean        4.070890
std         1.359663
min         1.000000
25%         4.000000
50%         5.000000
75%         5.000000
max         5.000000


### Amending the review_creation_date and review_answer_timestamp to datetime format

In [4]:
# amending to date time for column review_creation_date
orders_reviews_df['review_creation_date']=pd.to_datetime(orders_reviews_df['review_creation_date'])

# amending to date time for column review_answer_timestamp
orders_reviews_df['review_answer_timestamp']=pd.to_datetime(orders_reviews_df['review_answer_timestamp'])

print(orders_reviews_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 7 columns):
 #   Column                   Non-Null Count   Dtype         
---  ------                   --------------   -----         
 0   review_id                100000 non-null  object        
 1   order_id                 100000 non-null  object        
 2   review_score             100000 non-null  int64         
 3   review_comment_title     11715 non-null   object        
 4   review_comment_message   41753 non-null   object        
 5   review_creation_date     100000 non-null  datetime64[ns]
 6   review_answer_timestamp  100000 non-null  datetime64[ns]
dtypes: datetime64[ns](2), int64(1), object(4)
memory usage: 5.3+ MB
None


### Translating the review comment message

In [5]:
from deep_translator import GoogleTranslator
import pandas as pd
import numpy as np

# Function to translate text
def translate_text(text, dest_lang='en'):
    '''
    Function to translate text if it's not null, otherwise return the original text
    parameters: take in text to translate
    destinantion will be to english
    '''
    if pd.notnull(text):
        translation = GoogleTranslator(source='pt', target=dest_lang).translate(text)
        return translation
    else:
        return text   

    
    # # If the input is a list, concatenate the items into a single string
    # if isinstance(text, list):
    #     text = ' '.join(text)
    # return GoogleTranslator(source='pt', target=dest_lang).translate(text)


### Cleaning the dataframe column from special characters like \r \n

In [6]:
import re

# Function to remove special characters
def remove_special_characters(text):
    if pd.notnull(text):
        cleaned_text=text.replace('\r', '').replace('\n', '')
        return cleaned_text
    else:
        return text   
    
    
# Identify special characters in the 'Text' column
orders_reviews_df['cleaned_review_comment'] = orders_reviews_df['review_comment_message'].apply(remove_special_characters)

orders_reviews_df.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN           2018-01-18   
1                                                NaN           2018-03-10   
2                                                NaN           2018-02-17   
3              Recebi bem antes do prazo estipulado.           2017-04-21   
4  Parabéns lojas lannister adorei comprar pela I...           2018-03-01   

  review_answer_timestamp                             cleaned_review_comment  
0     2018-01-18 21:46:59                                                NaN  
1     2018-03-11 03:05:13                                                NaN  
2     2018-02-18 14:36:24                                                NaN  
3     2017-04-21 22:02:06              Recebi bem antes do prazo estipulado.  
4     2018-03-02 10:26:53  Parabéns lojas lannister adorei comprar pela I...

### Since the dataset is 41753 rows occured alot of premature termination therefore doing translation by each row

In [192]:
print(len(orders_reviews_df))

100000


### Translate from row 0 till 15900

In [ ]:
import time
batch_size=15000
total_rows = len(orders_reviews_df)

for idx, row in orders_reviews_df.iterrows():
    # idx=7000
    if idx < batch_size:
        # translate the text of each row of cleaned_review_comment
        translated_comment = translate_text(row['cleaned_review_comment'])
        
        # Update the DataFrame with translated texts
        orders_reviews_df.loc[idx, 'translated_review_comment'] = translated_comment
        
        #----------------------------------   
        #translate the review_comment_title
        translated_title = translate_text(row['review_comment_title'])
        # Update the DataFrame with translated texts
        orders_reviews_df.loc[idx, 'translated_comment_title'] = translated_title
    else:
        print(f"Translation complete for {idx} rows out of {len(orders_reviews_df)}")
        time.sleep(60) # wait for 3mins before proceeding the next 1000
        batch_size += 5000

print("All Translation complete")



orders_reviews_df.head()

### translating from row 15900 manage to translate till 95900 row

In [64]:
batch_size=20900
num_rows=len(orders_reviews_df)
for idx in range(15900, num_rows+1):
    if idx < batch_size:
        # translate the text of each row of cleaned_review_comment
        translated_comment = translate_text(orders_reviews_df.iloc[idx]['cleaned_review_comment'])

        # Update the DataFrame with translated texts
        orders_reviews_df.loc[idx, 'translated_review_comment'] = translated_comment

        #----------------------------------   
        #translate the review_comment_title
        translated_title = translate_text(row['review_comment_title'])

        # Update the DataFrame with translated texts
        orders_reviews_df.loc[idx, 'translated_comment_title'] = translated_title
    else:
        print(f"Translation complete for {idx} rows out of {len(orders_reviews_df)}")
        time.sleep(60) # wait for 3mins before proceeding the next 1000
        batch_size += 5000


Translation complete for 20900 rows out of 100000
Translation complete for 25900 rows out of 100000
Translation complete for 30900 rows out of 100000
Translation complete for 35900 rows out of 100000
Translation complete for 40900 rows out of 100000
Translation complete for 45900 rows out of 100000
Translation complete for 50900 rows out of 100000
Translation complete for 55900 rows out of 100000
Translation complete for 60900 rows out of 100000
Translation complete for 65900 rows out of 100000
Translation complete for 70900 rows out of 100000
Translation complete for 75900 rows out of 100000
Translation complete for 80900 rows out of 100000
Translation complete for 85900 rows out of 100000
Translation complete for 90900 rows out of 100000
Translation complete for 95900 rows out of 100000


IndexError: single positional indexer is out-of-bounds

In [70]:
orders_reviews_df.iloc[95890],['translated_review_comment']

(review_id                                     ab58357009a17bddaa7bbc410e04e955
 order_id                                      9fe32da158d0f7c1adb336ad3b69839b
 review_score                                                                 5
 review_comment_title                                              bom vendedor
 review_comment_message       Menor preço que achei entre lojas fisicas e vi...
 review_creation_date                                       2018-06-16 00:00:00
 review_answer_timestamp                                    2018-06-17 20:12:49
 cleaned_review_comment       Menor preço que achei entre lojas fisicas e vi...
 translated_review_comment    Lowest price I found between physical and virt...
 translated_comment_title                                                   NaN
 Name: 95890, dtype: object,
 ['translated_review_comment'])

translating from 95900 till 100000

In [73]:
for idx in range(95900, num_rows+1):
        # translate the text of each row of cleaned_review_comment
        translated_comment = translate_text(orders_reviews_df.iloc[idx]['cleaned_review_comment'])

        # Update the DataFrame with translated texts
        orders_reviews_df.loc[idx, 'translated_review_comment'] = translated_comment

        #----------------------------------   
        #translate the review_comment_title
        translated_title = translate_text(row['review_comment_title'])

        # Update the DataFrame with translated texts
        orders_reviews_df.loc[idx, 'translated_comment_title'] = translated_title
        print(f"Translation complete for {idx} rows out of {len(orders_reviews_df)}")

print(f"Completed translation")

Translation complete for 95900 rows out of 100000
Translation complete for 95901 rows out of 100000
Translation complete for 95902 rows out of 100000
Translation complete for 95903 rows out of 100000
Translation complete for 95904 rows out of 100000
Translation complete for 95905 rows out of 100000
Translation complete for 95906 rows out of 100000
Translation complete for 95907 rows out of 100000
Translation complete for 95908 rows out of 100000
Translation complete for 95909 rows out of 100000
Translation complete for 95910 rows out of 100000
Translation complete for 95911 rows out of 100000
Translation complete for 95912 rows out of 100000
Translation complete for 95913 rows out of 100000
Translation complete for 95914 rows out of 100000
Translation complete for 95915 rows out of 100000
Translation complete for 95916 rows out of 100000
Translation complete for 95917 rows out of 100000
Translation complete for 95918 rows out of 100000
Translation complete for 95919 rows out of 100000


IndexError: single positional indexer is out-of-bounds

In [74]:
orders_reviews_df.tail()

review_id                          order_id  \
99995  f3897127253a9592a73be9bdfdf4ed7a  22ec9f0669f784db00fa86d035cf8602   
99996  b3de70c89b1510c4cd3d0649fd302472  55d4004744368f5571d1f590031933e4   
99997  1adeb9d84d72fe4e337617733eb85149  7725825d039fc1f0ceb7635e3f7d9206   
99998  be360f18f5df1e0541061c87021e6d93  f8bd3f2000c28c5342fedeb5e50f2e75   
99999  efe49f1d6f951dd88b51e6ccd4cc548f  90531360ecb1eec2a1fbb265a0db0508   

       review_score review_comment_title  \
99995             5                  NaN   
99996             5                  NaN   
99997             4                  NaN   
99998             1                  NaN   
99999             1                  NaN   

                                  review_comment_message review_creation_date  \
99995                                                NaN           2017-12-09   
99996  Excelente mochila, entrega super rápida. Super...           2018-03-22   
99997                                                NaN           2018-07-01   
99998  Solicitei a compra de uma capa de retrovisor c...           2017-12-15   
99999  meu produto chegou e ja tenho que devolver, po...           2017-07-03   

      review_answer_timestamp  \
99995     2017-12-11 20:06:42   
99996     2018-03-23 09:10:43   
99997     2018-07-02 12:59:13   
99998     2017-12-16 01:29:43   
99999     2017-07-03 21:01:49   

                                  cleaned_review_comment  \
99995                                                NaN   
99996  Excelente mochila, entrega super rápida. Super...   
99997                                                NaN   
99998  Solicitei a compra de uma capa de retrovisor c...   
99999  meu produto chegou e ja tenho que devolver, po...   

                               translated_review_comment  \
99995                                                NaN   
99996  Excellent backpack, super fast delivery. I hig...   
99997                                                NaN   
99998  I requested the purchase of a Celtic/Prisma/Me...   
99999  My product arrived and I already have to retur...   

      translated_comment_title  
99995                      NaN  
99996                      NaN  
99997                      NaN  
99998                      NaN  
99999                      NaN

### Since we are only focussing on review comment message then we are dropping the rest and store it into a new dataframe

In [185]:
#drop the null values that are in review_comment_message and reset index and drop the old index column
text_analysis_df = orders_reviews_df.dropna(subset=['translated_review_comment']).reset_index(drop=True)

text_analysis_df.head()

#orders_reviews_df_no_nulls.info()

review_id                          order_id  \
0  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
1  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   
2  8670d52e15e00043ae7de4c01cc2fe06  b9bf720beb4ab3728760088589c62129   
3  4b49719c8a200003f700d3d986ea1a19  9d6f15f95d01e79bd1349cc208361f09   
4  3948b09f7c818e2d86c9a546758b2335  e51478e7e277a83743b6f9991dbfa3fb   

   review_score review_comment_title  \
0             5                  NaN   
1             5                  NaN   
2             4            recomendo   
3             4                  NaN   
4             5      Super recomendo   

                              review_comment_message review_creation_date  \
0              Recebi bem antes do prazo estipulado.           2017-04-21   
1  Parabéns lojas lannister adorei comprar pela I...           2018-03-01   
2  aparelho eficiente. no site a marca do aparelh...           2018-05-22   
3    Mas um pouco ,travando...pelo valor ta Boa.\r\n           2018-02-16   
4  Vendedor confiável, produto ok e entrega antes...           2018-05-23   

  review_answer_timestamp                             cleaned_review_comment  \
0     2017-04-21 22:02:06              Recebi bem antes do prazo estipulado.   
1     2018-03-02 10:26:53  Parabéns lojas lannister adorei comprar pela I...   
2     2018-05-23 16:45:47  aparelho eficiente. no site a marca do aparelh...   
3     2018-02-20 10:52:22        Mas um pouco ,travando...pelo valor ta Boa.   
4     2018-05-24 03:00:01  Vendedor confiável, produto ok e entrega antes...   

                           translated_review_comment translated_comment_title  
0  I received it well before the stipulated deadl...                      NaN  
1  Congratulations lannister stores I loved shopp...                      NaN  
2  efficient device. On the website the brand of ...              I recommend  
3      But a little slow...for the price, it's good.                      NaN  
4  Reliable seller, ok product and delivery on time.       I highly recommend

## Sentiment Analysis

In [ ]:
from textblob import TextBlob

# Function to perform sentiment analysis
def analyze_sentiment(text):
    blob = TextBlob(text)
    sentiment_score = blob.sentiment.polarity
    if sentiment_score > 0:
        return 'Positive'
    elif sentiment_score < 0:
        return 'Negative'
    else:
        return 'Neutral'

# Apply sentiment analysis to the 'Comment' column
text_analysis_df['Sentiment'] = text_analysis_df['translated_review_comment'].apply(analyze_sentiment)


or

## Keyword Extraction

import nltk
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from collections import Counter

# Download NLTK resources (run once)
nltk.download('punkt')
nltk.download('stopwords')

# Function to extract keywords from the text
def extract_keywords(text, num_keywords=5):
    tokens = word_tokenize(text.lower())
    stop_words = set(stopwords.words('english'))
    filtered_tokens = [word for word in tokens if word.isalnum() and word not in stop_words]
    keyword_counter = Counter(filtered_tokens)
    return keyword_counter.most_common(num_keywords)

# Apply keyword extraction to the 'Comment' column
text_analysis_df['Keywords'] = text_analysis_df['translated_review_comment'].apply(extract_keywords)


## Topic Modelling

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.decomposition import LatentDirichletAllocation

# Function to perform topic modeling using LDA
def perform_topic_modeling(texts, num_topics=5):
    vectorizer = CountVectorizer(max_df=0.95, min_df=2, stop_words='english')
    X = vectorizer.fit_transform(texts)
    lda = LatentDirichletAllocation(n_components=num_topics, random_state=42)
    lda.fit(X)
    return lda

# Apply topic modeling to the 'Comment' column
lda_model = perform_topic_modeling(text_analysis_df['translated_review_comment'])
